File: 101-symmetric-tree/symmetric-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSymmetric(root: TreeNode | null): boolean {
    return isMirror(root,root);
};

const isMirror = (leftNode : TreeNode , rightNode : TreeNode) => {
    if(leftNode === null) return rightNode === null;
    if(rightNode === null) return leftNode === null;
    if(leftNode.val !== rightNode.val) return false;
    const leftCheck = isMirror(leftNode.left , rightNode.right);
    const rightCheck = isMirror(leftNode.right , rightNode.left);
    return leftCheck && rightCheck;
}

File: 101-symmetric-tree/symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return isMirror(root.left , root.right);
    }

    private boolean isMirror(TreeNode leftChild , TreeNode rightChild) {
        if(leftChild == null && rightChild == null) return true;
        if(leftChild == null || rightChild == null) return false;
        boolean condition1 = leftChild.val == rightChild.val;
        boolean condition2 = isMirror(leftChild.left , rightChild.right);
        boolean condition3 = isMirror(leftChild.right , rightChild.left);
        return condition1 && (condition2 && condition3);
    }
}